app/controllers/annotations_controller.rb

```diff
@@ -10,46 +10,50 @@ def add_existing_annotation
     result = Result.find(params[:result_id])
     submission = result.submission
     submission_file = submission.submission_files.find(params[:submission_file_id])
+    begin
+      authorize! with: AnnotationPolicy
+      base_attributes = {
+        submission_file_id: submission_file.id,
+        is_remark: submission.has_remark?,
+        annotation_text_id: params[:annotation_text_id],
+        annotation_number: result.annotations.size + 1,
+        creator: current_user,
+        result_id: params[:result_id]
+      }
 
-    base_attributes = {
-      submission_file_id: submission_file.id,
-      is_remark: submission.has_remark?,
-      annotation_text_id: params[:annotation_text_id],
-      annotation_number: result.annotations.size + 1,
-      creator: current_user,
-      result_id: params[:result_id]
-    }
-
-    if submission_file.is_supported_image?
-      @annotation = result.annotations.create(
-        type: 'ImageAnnotation',
-        x1: params[:x1],
-        y1: params[:y1],
-        x2: params[:x2],
-        y2: params[:y2],
-        **base_attributes
-      )
-    elsif submission_file.is_pdf?
-      @annotation = result.annotations.create(
-        type: 'PdfAnnotation',
-        x1: params[:x1],
-        y1: params[:y1],
-        x2: params[:x2],
-        y2: params[:y2],
-        page: params[:page],
-        **base_attributes
-      )
-    else
-      @annotation = result.annotations.create(
-        type: 'TextAnnotation',
-        line_start: params[:line_start],
-        line_end: params[:line_end],
-        column_start: params[:column_start],
-        column_end: params[:column_end],
-        **base_attributes
-      )
+      if submission_file.is_supported_image?
+        @annotation = result.annotations.create(
+          type: 'ImageAnnotation',
+          x1: params[:x1],
+          y1: params[:y1],
+          x2: params[:x2],
+          y2: params[:y2],
+          **base_attributes
+        )
+      elsif submission_file.is_pdf?
+        @annotation = result.annotations.create(
+          type: 'PdfAnnotation',
+          x1: params[:x1],
+          y1: params[:y1],
+          x2: params[:x2],
+          y2: params[:y2],
+          page: params[:page],
+          **base_attributes
+        )
+      else
+        @annotation = result.annotations.create(
+          type: 'TextAnnotation',
+          line_start: params[:line_start],
+          line_end: params[:line_end],
+          column_start: params[:column_start],
+          column_end: params[:column_end],
+          **base_attributes
+        )
+      end
+      render :create
+    rescue ActionPolicy::Unauthorized => e
```

Comment:
rescuing from an `ActionPolicy::Unauthorized` error should go immediately after the call to `authorize!` if possible